homeassistant/auth/providers/ldap.py

```diff
@@ -0,0 +1,252 @@
+CONF_ALLOWED_GROUP_DNS = "allowed_group_dns"
+CONF_BASE_DN = "base_dn"
+CONF_BIND_TYPE = "bind_type"
+CONF_BIND_TYPE_ANONYMOUS = "anonymous"
+CONF_BIND_TYPE_AS_SERVICE_USER = "service-user"
+CONF_BIND_TYPE_AS_USER = "user"
+CONF_BIND_DN = "bind_dn"
+CONF_BIND_PASSWORD = "bind_password"
+CONF_CA_CERTS_FILE = "ca_certs_file"
+CONF_CERT_VALIDATION = "validate_certificates"
+CONF_ENCRYPTION = "encryption"
+CONF_ENCRYPTION_LDAPS = "ldaps"
+CONF_ENCRYPTION_NONE = "none"
+CONF_ENCRYPTION_STARTTLS = "starttls"
+CONF_PORT = "port"
+CONF_SERVER = "server"
+CONF_TIMEOUT = "timeout"
+CONF_USERNAME_ATTR = "username_attribute"
+
+# Default values
+DEFAULT_CONF_BIND_TYPE = CONF_BIND_TYPE_AS_SERVICE_USER
+DEFAULT_CONF_CERT_VALIDATION = True
+DEFAULT_CONF_PORT = 636
+DEFAULT_CONF_TIMEOUT = 10
+DEFAULT_CONF_USERNAME_ATTR = "uid"
+
+CONFIG_SCHEMA = AUTH_PROVIDER_SCHEMA.extend(
+    {
+        vol.Optional(CONF_ALLOWED_GROUP_DNS, default=[]): vol.All(
+            cv.ensure_list, [str]
+        ),
+        vol.Required(CONF_BASE_DN): str,
+        vol.Required(CONF_BIND_TYPE, default=DEFAULT_CONF_BIND_TYPE): vol.In(
+            [
+                CONF_BIND_TYPE_ANONYMOUS,
+                CONF_BIND_TYPE_AS_SERVICE_USER,
+                CONF_BIND_TYPE_AS_USER,
+            ]
+        ),
+        vol.Optional(CONF_BIND_DN): str,
+        vol.Optional(CONF_BIND_PASSWORD): str,
+        vol.Optional(CONF_CA_CERTS_FILE, default=None): str,
+        vol.Required(CONF_CERT_VALIDATION, default=DEFAULT_CONF_CERT_VALIDATION): bool,
+        vol.Required(CONF_ENCRYPTION, default=CONF_ENCRYPTION_LDAPS): vol.In(
+            [CONF_ENCRYPTION_LDAPS, CONF_ENCRYPTION_NONE, CONF_ENCRYPTION_STARTTLS]
+        ),
+        vol.Required(CONF_PORT, default=DEFAULT_CONF_PORT): int,
+        vol.Required(CONF_SERVER): str,
+        vol.Required(CONF_TIMEOUT, default=DEFAULT_CONF_TIMEOUT): int,
+        vol.Required(CONF_USERNAME_ATTR, default=DEFAULT_CONF_USERNAME_ATTR): str,
+    },
+    extra=vol.PREVENT_EXTRA,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class InvalidAuthError(HomeAssistantError):
+    """Raised when submitting invalid authentication."""
+
+
+class LdapError(HomeAssistantError):
+    """Raised when an LDAP error has been encountered."""
+
+
+@AUTH_PROVIDERS.register("ldap")
+class LdapAuthProvider(AuthProvider):
+    """LDAP auth provider."""
+
+    DEFAULT_TITLE = "LDAP Authentication"
+
+    async def async_login_flow(self, context: Optional[Dict]) -> LoginFlow:
+        """Return a flow to login."""
+        return LdapLoginFlow(self)
+
+    @callback
+    def async_validate_login(self, username: str, password: str) -> None:
+        """Validate a username and password."""
+        try:
+            tls = ldap3.Tls()
+            tls.validate = (
+                ssl.CERT_REQUIRED
+                if self.config[CONF_CERT_VALIDATION]
+                else ssl.CERT_NONE
+            )
+            tls.ca_certs_file = self.config[CONF_CA_CERTS_FILE]
```

Comment:
Relative paths from the configuration directory don't work right now. Use something like:
```
if self.config[CONF_CA_CERTS_FILE] is not None:
    tls.ca_certs_file = os.path.join(self.hass.config.path(), self.config[CONF_CA_CERTS_FILE])
```

Reply:
Hm. Is that the default behavior for Home Assistant? Do other components even support relative paths?

Reply:
Oops you're right!

Here is a proper example:

https://github.com/home-assistant/core/blob/508fc3fa0e083fdf1f6e05df332de404fc8a41a8/homeassistant/components/google_cloud/tts.py#L137-L142

Reply:
The following should work but I don't believe this type of logic is "normal" for Home Assistant:

```python
from pathlib import Path

if ca_certs_file:
    path = Path(ca_certs_file)
    if path.is_absolute():
        abs_path = path
    else:
        # Relative paths are relative to either cwd, or the config dir.
        for root_path in [
            Path.cwd(),
            Path(self.hass.config.path()),
        ]:
            tested_path = Path(root_path).joinpath(path)
            if tested_path.exists():
                abs_path = tested_path
                break

    tls.ca_certs_file = abs_path.resolve()
```

Reply:
The convention (and I am not sure where it is documented) is that paths are either absolute or relative to the config directory.

The following snipped would work for both absolute paths and relative to the config directory:

```
if self.config[CONF_CA_CERTS_FILE]:
    tls.ca_certs_file = self.hass.config.path(self.config[CONF_CA_CERTS_FILE])
     if not os.path.isfile(tls.ca_certs_file): 
         _LOGGER.error("File %s doesn't exist", tls.ca_certs_file) 
         return None
```

PS: Today I learned that `os.path.join("/config", "/a")` handles absolute directories properly and returns `/a`.

Reply:
Your new snippet won't work with absolute paths either.

`hass.config.path` **joins** the config path with the provided path arg.

https://github.com/home-assistant/core/blob/a92a7ec84848edfbb1187579ba4afcdc42b3d5a5/homeassistant/core.py#L1362-L1369

Reply:
Anyway. This should be fixed now -> 4a552a0

Reply:
As I was saying in my previous comment:

`os.path.join("/config", "/a")` returns `/a` so `hass.config.path()` works with absolute paths and paths relative to the configuration directory. `hass.config.path()` is what most components use to deal with paths.

But if your solution works for you, and the reviewers accept it I won't complain.

Reply:
Oh. I missed the PS. My sincerest apologies. I wasn't trying to be a dick about it. I just was looking for an excuse to use `pathlib` :-/

Reply:
Resolved in https://github.com/home-assistant/core/pull/37645/commits/7f30c22cdc4d2e13e192755e6ea8aa995ad05e8d

Reply:
Oh yes, I agree pathlib is great! With respect to the trivia I even made a [tweet](https://twitter.com/zeehio_/status/1288154040258494469?s=20) about it (I doubt it will reach more than a handful of people, though).

This is my story: I knew from using home assistant that absolute paths and paths relative to the config directory worked fine. I knew this was such a common problem that home assistant had to offer an easy function. Then I found the `hass.config.path()`, I saw its implementation and I thought exactly like you did. Absolute paths can't work with that! Then I checked some components and I saw they were just using that function. It was impossible! And then I tested the `os.path.join()` magic we just learned about. 🤯

I had to test it several times because I could not believe that behaviour, it is so well thought!

I think this PR is ready to be reviewed and (in my opinion) merged. You have put a lot of work. I have learned a lot of python and LDAP in the process. Thank you for your time.